bluemira/equilibria/coils.py

```diff
@@ -85,27 +92,318 @@ def generate_name(coil: Any, idx: Optional[int] = None):
+            # if dx or dz is not 0 and x,z inside coil
+            # TODO improve to remove inside coil calc if already known
+            inside = np.logical_and(self._points_inside_coil(x, z), not lg_or)
+
+            response = np.zeros(x.shape)
+
+            resp = response.reshape(-1)
+            ins = inside.reshape(-1)
+
+            if np.any(~inside):
+                resp[~ins] = greens_func(x[~inside], z[~inside]).reshape(-1)
+            if np.any(inside):
+                resp[ins] = semianalytic_func(x[inside], z[inside]).reshape(-1)
+        else:
+            response = greens_func(x, z)
+
+        return response
+
+    def _points_inside_coil(self, x, z):
+        """
+        Determine which points lie inside or on the coil boundary.
+
+        Parameters
+        ----------
+        x: Union[float, np.array]
+            The x coordinates to check
+        z: Union[float, np.array]
+            The z coordinates to check
+
+        Returns
+        -------
+        inside: np.array(dtype=bool)
+            The Boolean array of point indices inside/outside the coil boundary
+        """
+        x, z = np.ascontiguousarray(x), np.ascontiguousarray(z)
+        # Add an offset, to ensure points very near the edge are counted as
+        # being on the edge of a coil
+        atol = X_TOLERANCE
+        x_min, x_max = (
+            self._quad_x - self._quad_dx - atol,
+            self._quad_x + self._quad_dx + atol,
+        )
+        z_min, z_max = (
+            self._quad_z - self._quad_dz - atol,
+            self._quad_z + self._quad_dz + atol,
+        )
+        return (x >= x_min) & (x <= x_max) & (z >= z_min) & (z <= z_max)
+
+    def _control_Bx_greens(self, x, z):
+        """
+        Calculate radial magnetic field Bx respose at (x, z) due to a unit
+        current using Green's functions.
+        """
+        return np.add.reduceat(
+            greens_Bx(self._quad_x, self._quad_z, x, z) * self._quad_weighting,
+            self._no_quads,
+        )
+
+        # return sum(gx) / self.n_filaments
+
+    def _control_Bz_greens(self, x, z):
+        """
+        Calculate vertical magnetic field Bz at (x, z) due to a unit current
+        """
+        return np.add.reduceat(
+            greens_Bz(self._quad_x, self._quad_z, x, z) * self._quad_weighting,
+            self._no_quads,
+        )
+
+        # return sum(gz) / self.n_filaments
+
+    def _control_Bx_analytical(self, x, z):
+        """
+        Calculate radial magnetic field Bx response at (x, z) due to a unit
+        current using semi-analytic method.
+        """
+        return semianalytic_Bx(self.x, self.z, x, z, d_xc=self.dx, d_zc=self.dz)
+
+    def _control_Bz_analytical(self, x, z):
+        """
+        Calculate vertical magnetic field Bz response at (x, z) due to a unit
+        current using semi-analytic method.
+        """
+        return semianalytic_Bz(self.x, self.z, x, z, d_xc=self.dx, d_zc=self.dz)
```

Comment:
x, and z don't exist anymore

Reply:
this was taken over from the old coils because it made no distinction to subcoils so I assumed it was the same calculation everywhere.
Also see below comment about existence